File: src/python/surface/arbitrage.py
"""
Arbitrage Checking Framework
Ensures IV surfaces are arbitrage-free before deployment

Key arbitrage conditions:
1. Butterfly: Call prices must be convex in strike (∂²C/∂K² ≥ 0)
2. Calendar: Call prices must be non-decreasing in time (∂C/∂T ≥ 0)
3. Total Variance: w = σ²T must be non-decreasing in maturity

**OPTIMIZED:** All computations delegated to C++ engine
"""

from dataclasses import dataclass
from typing import List, Optional, Tuple
import numpy as np
import logging

logger = logging.getLogger(__name__)

try:
    from ..cpp_unified_engine import get_unified_cpp_engine
except ImportError:
    from cpp_unified_engine import get_unified_cpp_engine


@dataclass
class ArbitrageViolation:
    """Single arbitrage violation"""
    type: str  # 'butterfly', 'calendar', 'total_variance'
    severity: str  # 'minor', 'moderate', 'severe'
    value: float  # Violation magnitude
    tolerance: float  # Threshold used
    location: Tuple  # (strike/maturity context)
    message: str


@dataclass
class ArbitrageReport:
    """Complete arbitrage validation report"""
    is_arbitrage_free: bool
    violations: List[ArbitrageViolation]
    butterfly_violations: int
    calendar_violations: int
    total_variance_violations: int
    summary: str


class ArbitrageChecker:
    """
    Arbitrage-free surface validator using C++ engine
    
    Investment bank standards:
    - Zero tolerance for severe violations (>5x threshold)
    - Bid-ask spread accommodation for minor violations
    - Comprehensive reporting with severity classification
    
    **C++ OPTIMIZED:** All computational operations delegated to C++ engine
    """
    
    def __init__(self, tolerance: float = 1e-6, bid_ask_buffer: float = 0.0):
        """
        Initialize arbitrage checker
        
        Args:
            tolerance: Base tolerance for violations (default 1e-6)
            bid_ask_buffer: Additional buffer for bid-ask spreads
        """
        self.tolerance = tolerance
        self.bid_ask_buffer = bid_ask_buffer
        self.engine = get_unified_cpp_engine()
        if self.engine is None:
            raise RuntimeError("C++ engine unavailable — ArbitrageChecker requires C++ engine")
    
    @staticmethod
    def _classify_severity(violation_magnitude: float, tolerance: float) -> str:
        """Classify violation severity based on magnitude relative to tolerance"""
        if tolerance <= 0:
            return 'severe'
        
        ratio = violation_magnitude / tolerance
        if ratio < 2.0:
            return 'minor'
        elif ratio < 5.0:
            return 'moderate'
        else:
            return 'severe'
    
    def check_butterfly_arbitrage(
        self,
        strikes: np.ndarray,
        implied_vols: np.ndarray,
        maturity: float,
        spot: float,
        rate: float = 0.0,
        dividend_yield: float = 0.0,
        bid_ask_spreads: Optional[np.ndarray] = None
    ) -> Tuple[bool, List[ArbitrageViolation]]:
        """
        Check butterfly arbitrage: call prices must be convex in strike
        
        Condition: C(K₁) - 2C(K₂) + C(K₃) ≥ 0 for K₁ < K₂ < K₃
        Equivalently: ∂²C/∂K² ≥ 0
        
        Args:
            strikes: Array of strike prices (sorted ascending)
            implied_vols: Array of implied volatilities
            maturity: Time to maturity (years)
            spot: Spot price
            rate: Risk-free rate
            dividend_yield: Dividend yield
            bid_ask_spreads: Optional bid-ask spreads per strike
            
        Returns:
            (is_valid, violations): Tuple of validation result and list of violations
        """
        # Adjust tolerance to accommodate bid-ask spreads if provided
        tol = self.tolerance
        if bid_ask_spreads is not None:
            tol = float(np.mean(np.maximum(tol, 0.5 * bid_ask_spreads + self.bid_ask_buffer)))

        call_prices = self.engine.black_scholes_prices(
            spot, strikes, maturity, rate, implied_vols, dividend_yield
        )
        violation_count = self.engine.check_butterfly_arbitrage(
            strikes, call_prices, tol
        )
        if violation_count == 0:
            return True, []
        violations = [
            ArbitrageViolation(
                type='butterfly',
                severity='moderate',
                value=-tol,
                tolerance=tol,
                location=(float(strikes[0]), float(strikes[-1]), maturity),
                message=f"Butterfly arbitrage: {violation_count} violation(s) at T={maturity:.3f}"
            )
            for _ in range(violation_count)
        ]
        return False, violations
    
    def check_calendar_arbitrage(
        self,
        strike: float,
        maturities: np.ndarray,
        ivs: np.ndarray,
        spot: float,
        rate: float = 0.0,
        dividend_yield: float = 0.0,
        bid_ask_spreads: Optional[np.ndarray] = None
    ) -> Tuple[bool, List[ArbitrageViolation]]:
        """
        Check calendar arbitrage: call prices must be non-decreasing in time
        
        Condition: C(T₁) ≤ C(T₂) for T₁ < T₂ at fixed strike
        Equivalently: ∂C/∂T ≥ 0
        
        Args:
            strike: Strike price
            maturities: Array of maturities (sorted ascending)
            ivs: Array of implied volatilities
            spot: Spot price
            rate: Risk-free rate
            dividend_yield: Dividend yield
            bid_ask_spreads: Optional bid-ask spreads per maturity
            
        Returns:
            (is_valid, violations): Tuple of validation result and list of violations
        """
        # Adjust tolerance to accommodate bid-ask spreads if provided
        tol = self.tolerance
        if bid_ask_spreads is not None:
            tol = float(np.mean(np.maximum(tol, 0.5 * bid_ask_spreads + self.bid_ask_buffer)))

        # bs_prices takes a scalar maturity; compute one price per maturity point
        call_prices = np.array([
            float(self.engine.black_scholes_prices(
                spot, np.array([strike]), float(mat), rate, np.array([iv]), dividend_yield
            )[0])
            for mat, iv in zip(maturities, ivs)
        ])
        violation_count = self.engine.check_calendar_arbitrage(
            maturities, call_prices, tol
        )
        if violation_count == 0:
            return True, []
        violations = [
            ArbitrageViolation(
                type='calendar',
                severity='moderate',
                value=-tol,
                tolerance=tol,
                location=(strike, float(maturities[0]), float(maturities[-1])),
                message=f"Calendar arbitrage: {violation_count} violation(s) at K={strike:.2f}"
            )
            for _ in range(violation_count)
        ]
        return False, violations
    
    def check_total_variance_monotonicity(
        self,
        strike: float,
        maturities: np.ndarray,
        ivs: np.ndarray
    ) -> Tuple[bool, List[ArbitrageViolation]]:
        """
        Check total variance monotonicity: w = σ²T must be non-decreasing
        
        Condition: σ₁²T₁ ≤ σ₂²T₂ for T₁ < T₂
        This is a stronger condition than calendar arbitrage
        
        Args:
            strike: Strike price
            maturities: Array of maturities (sorted ascending)
            ivs: Array of implied volatilities
            
        Returns:
            (is_valid, violations): Tuple of validation result and list of violations
        """
        if len(maturities) < 2:
            return True, []
        
        total_variances = ivs**2 * maturities
        violations = []
        
        for i in range(len(maturities) - 1):
            T1, T2 = maturities[i], maturities[i+1]
            w1, w2 = total_variances[i], total_variances[i+1]
            
            if w2 < w1 - self.tolerance:
                severity = self._classify_severity(abs(w2 - w1), self.tolerance)
                violations.append(ArbitrageViolation(
                    type='total_variance',
                    severity=severity,
                    value=w2 - w1,
                    tolerance=self.tolerance,
                    location=(strike, T1, T2),
                    message=f"Total variance non-monotonic at K={strike:.2f}, T=({T1:.3f}, {T2:.3f}): w₂ - w₁ = {w2 - w1:.6f} < {-self.tolerance:.6f}"
                ))
        
        return len(violations) == 0, violations
    
    def validate_surface(
        self,
        strikes: np.ndarray,
        maturities: np.ndarray,
        implied_vols: np.ndarray,
        spot: float,
        rate: float = 0.0,
        dividend_yield: float = 0.0,
        bid_ask_spreads: Optional[np.ndarray] = None
    ) -> ArbitrageReport:
        """
        Complete arbitrage validation of IV surface
        
        Checks:
        1. Butterfly arbitrage across all maturities
        2. Calendar arbitrage across all strikes
        3. Total variance monotonicity
        
        Args:
            strikes: 1D array of strikes
            maturities: 1D array of maturities
            implied_vols: 2D array (strikes × maturities) of IVs
            spot: Spot price
            rate: Risk-free rate
            dividend_yield: Dividend yield
            bid_ask_spreads: Optional 2D array of bid-ask spreads
            
        Returns:
            ArbitrageReport with comprehensive results
        """
        all_violations = []
        
        # 1. Butterfly checks (for each maturity slice)
        butterfly_count = 0
        for j, T in enumerate(maturities):
            ivs_at_T = implied_vols[:, j]
            spreads_at_T = bid_ask_spreads[:, j] if bid_ask_spreads is not None else None
            
            is_valid, violations = self.check_butterfly_arbitrage(
                strikes, ivs_at_T, T, spot, rate, dividend_yield, spreads_at_T
            )
            butterfly_count += len(violations)
            all_violations.extend(violations)
        
        # 2. Calendar checks (for each strike slice)
        calendar_count = 0
        for i, K in enumerate(strikes):
            ivs_at_K = implied_vols[i, :]
            spreads_at_K = bid_ask_spreads[i, :] if bid_ask_spreads is not None else None
            
            is_valid, violations = self.check_calendar_arbitrage(
                K, maturities, ivs_at_K, spot, rate, dividend_yield, spreads_at_K
            )
            calendar_count += len(violations)
            all_violations.extend(violations)
        
        # 3. Total variance checks
        tv_count = 0
        for i, K in enumerate(strikes):
            ivs_at_K = implied_vols[i, :]
            is_valid, violations = self.check_total_variance_monotonicity(
                K, maturities, ivs_at_K
            )
            tv_count += len(violations)
            all_violations.extend(violations)
        
        # Generate report
        is_arbitrage_free = len(all_violations) == 0
        
        summary_lines = [
            f"Arbitrage Check Results:",
            f"  Butterfly violations: {butterfly_count}",
            f"  Calendar violations: {calendar_count}",
            f"  Total variance violations: {tv_count}",
            f"  Status: {'✓ PASS - Arbitrage-free' if is_arbitrage_free else '✗ FAIL - Arbitrage detected'}"
        ]
        
        if all_violations:
            severe = sum(1 for v in all_violations if v.severity == 'severe')
            moderate = sum(1 for v in all_violations if v.severity == 'moderate')
            minor = sum(1 for v in all_violations if v.severity == 'minor')
            summary_lines.append(f"  Severity: {severe} severe, {moderate} moderate, {minor} minor")
        
        return ArbitrageReport(
            is_arbitrage_free=is_arbitrage_free,
            violations=all_violations,
            butterfly_violations=butterfly_count,
            calendar_violations=calendar_count,
            total_variance_violations=tv_count,
            summary='\n'.join(summary_lines)
        )

File: src/cpp/src/arbitrage.cpp
#include "arbitrage.hpp"
#include "pricing.hpp"
#include <algorithm>
#include <cmath>
#include <limits>

namespace iv_surface
{

    std::string ArbitrageChecker::classify_severity(double violation_magnitude, double tolerance)
    {
        if (tolerance <= 0.0)
        {
            return "severe";
        }

        double ratio = violation_magnitude / tolerance;

        if (ratio < 2.0)
        {
            return "minor";
        }
        else if (ratio < 5.0)
        {
            return "moderate";
        }
        else
        {
            return "severe";
        }
    }

    std::vector<ArbitrageChecker::Violation> ArbitrageChecker::check_butterfly(
        const std::vector<double> &strikes,
        const std::vector<double> &call_prices,
        double maturity,
        double tolerance,
        const std::vector<double> *bid_ask_spreads)
    {
        std::vector<Violation> violations;

        if (strikes.size() < 3 || call_prices.size() < 3)
        {
            return violations;
        }

        const std::size_t n = std::min(strikes.size(), call_prices.size());

        // Check each consecutive triplet
        for (std::size_t i = 1; i + 1 < n; ++i)
        {
            const double K_left = strikes[i - 1];
            const double K_mid = strikes[i];
            const double K_right = strikes[i + 1];

            const double C_left = call_prices[i - 1];
            const double C_mid = call_prices[i];
            const double C_right = call_prices[i + 1];

            const double h_left = K_mid - K_left;
            const double h_right = K_right - K_mid;

            if (h_left <= 0.0 || h_right <= 0.0)
            {
                continue;
            }

            // Butterfly value
            double butterfly;
            if (std::abs(h_left - h_right) < 1e-6)
            {
                // Equal spacing: standard butterfly
                butterfly = C_left - 2.0 * C_mid + C_right;
            }
            else
            {
                // Unequal spacing: linear interpolation check
                const double alpha = h_right / (h_left + h_right);
                const double C_mid_linear = alpha * C_left + (1.0 - alpha) * C_right;
                butterfly = C_mid_linear - C_mid;
            }

            // Determine tolerance
            double tol = tolerance;
            if (bid_ask_spreads != nullptr && bid_ask_spreads->size() > i)
            {
                tol = 0.5 * (*bid_ask_spreads)[i];
            }

            // Check violation
            if (butterfly < -tol)
            {
                Violation v;
                v.type = "butterfly";
                v.value = butterfly;
                v.tolerance = tol;
                v.location = {K_left, K_mid, K_right, maturity};
                v.severity = classify_severity(std::abs(butterfly), tol);
                violations.push_back(v);
            }
        }

        return violations;
    }

    std::vector<ArbitrageChecker::Violation> ArbitrageChecker::check_calendar(
        double strike,
        const std::vector<double> &maturities,
        const std::vector<double> &call_prices,
        double tolerance,
        const std::vector<double> *bid_ask_spreads)
    {
        std::vector<Violation> violations;

        if (maturities.size() < 2 || call_prices.size() < 2)
        {
            return violations;
        }

        const std::size_t n = std::min(maturities.size(), call_prices.size());

        // Check each consecutive pair
        for (std::size_t i = 0; i + 1 < n; ++i)
        {
            const double T1 = maturities[i];
            const double T2 = maturities[i + 1];
            const double C1 = call_prices[i];
            const double C2 = call_prices[i + 1];

            // Calendar spread value (should be non-negative)
            const double calendar = C2 - C1;

            // Determine tolerance
            double tol = tolerance;
            if (bid_ask_spreads != nullptr && bid_ask_spreads->size() > i + 1)
            {
                tol = 0.5 * ((*bid_ask_spreads)[i] + (*bid_ask_spreads)[i + 1]);
            }

            // Check violation
            if (calendar < -tol)
            {
                Violation v;
                v.type = "calendar";
                v.value = calendar;
                v.tolerance = tol;
                v.location = {strike, T1, T2};
                v.severity = classify_severity(std::abs(calendar), tol);
                violations.push_back(v);
            }
        }

        return violations;
    }

    std::vector<ArbitrageChecker::Violation> ArbitrageChecker::check_total_variance(
        const std::vector<double> &strikes,
        const std::vector<double> &maturities,
        const std::vector<double> &implied_vols,
        double tolerance)
    {
        std::vector<Violation> violations;

        if (strikes.empty() || maturities.empty())
        {
            return violations;
        }

        const std::size_t n_strikes = strikes.size();
        const std::size_t n_maturities = maturities.size();

        if (implied_vols.size() != n_strikes * n_maturities)
        {
            return violations;
        }

        // Check monotonicity for each strike
        for (std::size_t i = 0; i < n_strikes; ++i)
        {
            const double K = strikes[i];

            // Check each consecutive maturity pair
            for (std::size_t j = 0; j + 1 < n_maturities; ++j)
            {
                const double T1 = maturities[j];
                const double T2 = maturities[j + 1];

                // Access IVs (row-major indexing)
                const double sigma1 = implied_vols[i * n_maturities + j];
                const double sigma2 = implied_vols[i * n_maturities + j + 1];

                if (!std::isfinite(sigma1) || !std::isfinite(sigma2))
                {
                    continue;
                }

                // Compute total variance w = σ²T
                const double w1 = (T1 > 0.0) ? sigma1 * sigma1 * T1 : 0.0;
                const double w2 = (T2 > 0.0) ? sigma2 * sigma2 * T2 : 0.0;

                // Should have w2 >= w1
                const double diff = w2 - w1;

                // Tolerance scales with magnitude
                const double tol = std::max(tolerance, tolerance * std::abs(w1));

                if (diff < -tol)
                {
                    Violation v;
                    v.type = "total_variance";
                    v.value = diff;
                    v.tolerance = tol;
                    v.location = {K, T1, T2};
                    v.severity = classify_severity(std::abs(diff), tol);
                    violations.push_back(v);
                }
            }
        }

        return violations;
    }

    ArbitrageChecker::Report ArbitrageChecker::validate_surface(
        const std::vector<double> &strikes,
        const std::vector<double> &maturities,
        const std::vector<double> &implied_vols,
        double spot,
        double rate,
        double dividend_yield,
        double tolerance,
        const std::vector<double> *bid_ask_spreads)
    {
        Report report;
        report.passed = true;
        report.total_checks = 0;
        report.butterfly_pass = true;
        report.calendar_pass = true;
        report.total_variance_pass = true;

        const std::size_t n_strikes = strikes.size();
        const std::size_t n_maturities = maturities.size();

        if (implied_vols.size() != n_strikes * n_maturities)
        {
            return report;
        }

        // 1. Butterfly checks (for each maturity)
        for (std::size_t j = 0; j < n_maturities; ++j)
        {
            const double T = maturities[j];
            if (T <= 0.0)
            {
                continue;
            }

            // Convert IVs to call prices
            std::vector<double> call_prices(n_strikes);
            for (std::size_t i = 0; i < n_strikes; ++i)
            {
                const double K = strikes[i];
                const double sigma = implied_vols[i * n_maturities + j];

                if (sigma > 0.0 && std::isfinite(sigma))
                {
                    call_prices[i] = BlackScholesEngine::call_price(
                        spot, K, T, rate, sigma, dividend_yield);
                }
                else
                {
                    call_prices[i] = 0.0;
                }
            }

            // Bid-ask spreads for this maturity (if provided)
            const std::vector<double> *spreads_at_T = nullptr;
            std::vector<double> spreads_slice;
            if (bid_ask_spreads != nullptr)
            {
                spreads_slice.resize(n_strikes);
                for (std::size_t i = 0; i < n_strikes; ++i)
                {
                    spreads_slice[i] = (*bid_ask_spreads)[i * n_maturities + j];
                }
                spreads_at_T = &spreads_slice;
            }

            auto viols = check_butterfly(strikes, call_prices, T, tolerance, spreads_at_T);
            report.total_checks += static_cast<int>(std::max<std::size_t>(1, n_strikes - 2));
            if (!viols.empty())
            {
                report.butterfly_pass = false;
                report.violations.insert(report.violations.end(), viols.begin(), viols.end());
            }
        }

        // 2. Calendar checks (for each strike)
        for (std::size_t i = 0; i < n_strikes; ++i)
        {
            const double K = strikes[i];

            // Extract call prices at this strike
            std::vector<double> call_prices(n_maturities);
            for (std::size_t j = 0; j < n_maturities; ++j)
            {
                const double T = maturities[j];
                const double sigma = implied_vols[i * n_maturities + j];

                if (T > 0.0 && sigma > 0.0 && std::isfinite(sigma))
                {
                    call_prices[j] = BlackScholesEngine::call_price(
                        spot, K, T, rate, sigma, dividend_yield);
                }
                else
                {
                    call_prices[j] = 0.0;
                }
            }

            // Bid-ask spreads for this strike (if provided)
            const std::vector<double> *spreads_at_K = nullptr;
            std::vector<double> spreads_slice;
            if (bid_ask_spreads != nullptr)
            {
                spreads_slice.resize(n_maturities);
                for (std::size_t j = 0; j < n_maturities; ++j)
                {
                    spreads_slice[j] = (*bid_ask_spreads)[i * n_maturities + j];
                }
                spreads_at_K = &spreads_slice;
            }

            auto viols = check_calendar(K, maturities, call_prices, tolerance, spreads_at_K);
            report.total_checks += static_cast<int>(std::max<std::size_t>(1, n_maturities - 1));
            if (!viols.empty())
            {
                report.calendar_pass = false;
                report.violations.insert(report.violations.end(), viols.begin(), viols.end());
            }
        }

        // 3. Total variance monotonicity
        auto viols = check_total_variance(strikes, maturities, implied_vols, tolerance);
        report.total_checks += static_cast<int>(n_strikes * std::max<std::size_t>(1, n_maturities - 1));
        if (!viols.empty())
        {
            report.total_variance_pass = false;
            report.violations.insert(report.violations.end(), viols.begin(), viols.end());
        }

        // Final verdict
        report.passed = report.butterfly_pass && report.calendar_pass && report.total_variance_pass;

        return report;
    }

} // namespace iv_surface
